### PR TITLE
Interpreter refactor

### PR DESCRIPTION
**Pull request by XiaowenHu96**
This PR refactors the interpreter.

1. Implement super-instructions for indexed/project/existence operations. 
We specialize `Constant` and `TupleElements` so that no virtual dispatch is needed for them during interpreter execution.
~~This supposes to bring about 10% speedup on the interpreter based on some previous benchmark result with an older version of Souffle, there is an ongoing performance benchmarking running right now for latest data.~~
Update: This brings the interpreter 10-15% speedup due to reduction in dispatch.

2. With this PR, the interpreter does not emit `TupleOperation` node during code generation unless the profile option is turned on. This will further save us 10-15% dispatches.

3. Refactor Interpreter class. 
Before this PR, the interpreter refers to the operation arguments simply by their index positions. Such an approach can be hard to read and debug.  This PR formalizing the interpreter class hierarchy by assigning each interpreter operation with its own class. Then by inheriting helper classes, the interpreter can refer to its operation arguments with more meaningful names.

### CODE REVIEW

**Comment by b-scholz** [comment]
Can you do a clang-format -i *.cpp *.h in the interpreter directory? The format is not quite right.

File: src/interpreter/InterpreterNode.h [hunk 1]
```diff
@@ -132,23 +257,518 @@ class InterpreterNode {
...
+
+/**
+ * @class InterpreterCondOperation
+ * @brief Encode a conditional operation for the interpreter node. E.g. Exit, Filter
+ */
+class InterpreterCondOperation {
+public:
+    InterpreterCondOperation(std::unique_ptr<InterpreterNode> cond) : cond(std::move(cond)) {}
+
+    inline const InterpreterNode* getCondtion() const {
```
Comment (mmcgr) [inline src/interpreter/InterpreterNode.h:327]:
```suggestion
    inline const InterpreterNode* getCondition() const {
```

File: src/interpreter/InterpreterNode.h [hunk 2]
```diff
@@ -112,16 +111,142 @@ class InterpreterNode {
...
+    std::unique_ptr<InterpreterNode> lhs;
+    std::unique_ptr<InterpreterNode> rhs;
+};
+
+/**
+ * @class InterpreterSuperInst
+ * @brief This class encodes information for a super-instruction, which is
+ *        used to eliminate Number and TupleElement in index/project/existence operation.
+ */
+class InterpreterSuperInst {
```
Comment (mmcgr) [inline src/interpreter/InterpreterNode.h:201]: `Inst` can be short for many things, including related things. Use the full word for clarity:
```suggestion
class InterpreterSuperInstruction {
```

File: src/interpreter/InterpreterNode.h [hunk 3]
```diff
@@ -132,23 +257,518 @@ class InterpreterNode {
...
+
+protected:
+    std::unique_ptr<InterpreterNode> nested;
+};
+
+/**
+ * @class InterpreterCondOperation
+ * @brief Encode a conditional operation for the interpreter node. E.g. Exit, Filter
+ */
+class InterpreterCondOperation {
```
Comment (mmcgr) [inline src/interpreter/InterpreterNode.h:323]: You could use the full word here.
```suggestion
class InterpreterConditionalOperation {
```

File: src/interpreter/InterpreterNode.h [hunk 4]
```diff
@@ -112,16 +111,142 @@ class InterpreterNode {
...
+    enum InterpreterNodeType type;
+    const RamNode* shadow;
+    RelationHandle* const relHandle;
+};
+
+/**
+ * @class InterpreterCompundNode
+ * @brief Compound node represents interpreter node with a list of children.
+ */
+class InterpreterCompundNode : public InterpreterNode {
```
Comment (mmcgr) [inline src/interpreter/InterpreterNode.h:130]:
```suggestion
class InterpreterCompoundNode : public InterpreterNode {
```

File: src/interpreter/InterpreterNode.h [hunk 5]
```diff
@@ -132,23 +257,518 @@ class InterpreterNode {
...
+
+/**
+ * @class InterpreterNestedOperation
+ * @brief Encode a nested operation for the interpreter node. E.g. Loop, IndexScan
+ */
+class InterpreterNestedOperation {
+public:
+    InterpreterNestedOperation(std::unique_ptr<InterpreterNode> nested) : nested(std::move(nested)) {}
+
+    inline const InterpreterNode* getNestesOperation() const {
```
Comment (mmcgr) [inline src/interpreter/InterpreterNode.h:311]:
```suggestion
    inline const InterpreterNode* getNestedOperation() const {
```

**Comment by XiaowenHu96** [comment]
> A few typos and naming changes

Thanks for checking that for me. Fixed.

**Pull request merged**